src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,728 @@
+                             self._left_finger_id,
+                             physicsClientId=self._physics_client_id)[0]
+        state.set(self._robot, "pose_x", rx)
+        state.set(self._robot, "pose_y", ry)
+        state.set(self._robot, "pose_z", rz)
+        state.set(self._robot, "fingers", rf)
+
+        # Get block states.
+        for block_id, block in self._block_id_to_block.items():
+            (bx, by, bz), _ = p.getBasePositionAndOrientation(
+                block_id, physicsClientId=self._physics_client_id)
+            held = (block_id == self._held_block_id)
+            state.set(block, "pose_x", bx)
+            state.set(block, "pose_y", by)
+            state.set(block, "pose_z", bz)
+            state.set(block, "held", held)
+
+        return state
+
+    def _detect_held_block(self) -> Optional[int]:
+        """Return the PyBullet object ID of the held object if one exists.
+
+        If multiple blocks are within the grasp tolerance, return the
+        one that is closest.
+        """
+        expected_finger_normals = {
+            self._left_finger_id: np.array([0., 1., 0.]),
+            self._right_finger_id: np.array([0., -1., 0.]),
+        }
+        closest_held_block = None
+        closest_held_block_dist = float("inf")
+        for block_id in self._block_id_to_block:
+            for finger_id, expected_normal in expected_finger_normals.items():
+                # Find points on the block that are within grasp_tol distance
+                # of the finger.
+                closest_points = p.getClosestPoints(
+                    bodyA=self._fetch_id,
+                    bodyB=block_id,
+                    distance=self._grasp_tol,
+                    linkIndexA=finger_id,
+                    physicsClientId=self._physics_client_id)
+                for point in closest_points:
+                    # If the contact normal is substantially different from
+                    # the expected contact normal, this is probably a block
+                    # on the outside of the gripper, rather than the inside.
+                    # A perfect score here is 1.0 (normals are unit vectors).
+                    contact_normal = point[7]
+                    if expected_normal.dot(contact_normal) < 0.5:
+                        continue
+                    # Handle the case where multiple blocks pass this check
+                    # by taking the closest one. This should be rare, but it
+                    # can happen when two blocks are stacked and the robot is
+                    # unstacking the top one.
+                    contact_distance = point[8]
+                    if contact_distance < closest_held_block_dist:
+                        closest_held_block = block_id
+                        closest_held_block_dist = contact_distance
+        return closest_held_block
+
+    ########################## Parameterized Options ##########################
+
+    def _move_to_pose(
```

Comment:
it looks like this and other options are specific to *gripper* poses. we should clarify that by sprinkling in the word "gripper" throughout, in both the method names and the docstrings, as appropriate.

Reply:
I'm reserving the word "gripper" to mean "the left and right fingers". I'll sprinkle the word "end effector"

Reply:
actually I'll just use fingers and end effector and purge gripper